packages/coinjoin/src/client/round/outputDecomposition.ts

```diff
@@ -0,0 +1,212 @@
+        outputSize,
+    };
+};
+
+export const getOutputAmounts = async (
+    round: CoinjoinRound,
+    accountKey: string,
+    availableVsize: number,
+    outputSize: number,
+    options: CoinjoinRoundOptions,
+) => {
+    const { roundParameters } = round;
+    const { signal, middlewareUrl } = options;
+
+    const registeredInputs = getRoundEvents('InputAdded', round.coinjoinState.events);
+    const internalAmounts: number[] = [];
+    const externalAmounts: number[] = [];
+
+    registeredInputs.forEach(i => {
+        const internal = round.inputs.find(
+            input =>
+                input.accountKey === accountKey && compareOutpoint(input.outpoint, i.coin.outpoint),
+        );
+        if (internal) {
+            internalAmounts.push(internal.confirmedAmountCredentials![0].value);
+        } else {
+            const size = getExternalOutputSize(i.coin.txOut.scriptPubKey);
+            const miningFee = Math.floor((size * roundParameters.miningFeeRate) / 1000);
+            const coordinatorFee = Math.floor(
+                roundParameters.coordinationFeeRate.rate * i.coin.txOut.value,
+            );
+            externalAmounts.push(i.coin.txOut.value - coordinatorFee - miningFee);
+        }
+    });
+    const outputAmounts = await middleware.decomposeAmounts(
+        {
+            feeRate: roundParameters.miningFeeRate,
+            allowedOutputAmounts: roundParameters.allowedOutputAmounts,
+        },
+        outputSize,
+        availableVsize, // TODO: sum all available vsize and use chunks in output registration (increase nr of outputs)
+        internalAmounts,
+        externalAmounts,
+        { signal, baseUrl: middlewareUrl },
+    );
+    options.log(`Decompose amounts: ${outputAmounts.join('')}`);
+    return outputAmounts.map(amount => {
+        const miningFee = Math.floor((outputSize * roundParameters.miningFeeRate) / 1000);
+        const coordinatorFee = 0; // TODO: should be Math.floor(roundParameters.coordinationFeeRate.rate * amount);
+        return amount + coordinatorFee + miningFee; // TODO: should be plebs dont pay?
+    });
+};
+
+export interface DecomposedOutputs {
+    accountKey: string;
+    outputSize: number;
+    amounts: number[];
+    amountCredentials: middleware.Credentials[];
+    vsizeCredentials: middleware.Credentials[];
+}
+
+export const outputDecomposition = async (
+    round: CoinjoinRound,
+    options: CoinjoinRoundOptions,
+): Promise<DecomposedOutputs[]> => {
+    // group inputs by accountKeys
+    const groupInputsByAccount = round.inputs.reduce((result, input) => {
+        if (!result[input.accountKey]) {
+            result[input.accountKey] = [];
+        }
+        if (!input.confirmedAmountCredentials || !input.confirmedVsizeCredentials) {
+            throw new Error(`Missing confirmed credentials for ${input.outpoint}`);
+        }
+        result[input.accountKey].push(input);
+        return result;
+    }, {} as Record<string, Alice[]>);
```

Comment:
nit: I think this way of typing `reduce` is slightly better (you don't need the type assertion)
```suggestion
    const groupInputsByAccount = round.inputs.reduce<Record<string, Alice[]>>((result, input) => {
        if (!result[input.accountKey]) {
            result[input.accountKey] = [];
        }
        if (!input.confirmedAmountCredentials || !input.confirmedVsizeCredentials) {
            throw new Error(`Missing confirmed credentials for ${input.outpoint}`);
        }
        result[input.accountKey].push(input);
        return result;
    }, {});
```